TESTS/mbed_drivers/mem_trace/main.cpp

```diff
@@ -110,180 +126,305 @@ extern "C" void test_trace_cb(uint8_t op, void *res, void *caller, ...) {
+{
+    volatile  static int trace_guard = 0;
+    trace_guard++;
+    TEST_ASSERT_TRUE_MESSAGE(trace_guard == 1, "Race condition occurred !!!!");
+    trace_guard--;
+}
 
-// Allocate a single buffer, then free it. Check that tracing matches the operations.
-static void test_case_single_malloc_free() {
+// Thread function
+void malloc_free(volatile bool *thread_continue)
+{
+    const size_t block_size = 126;
+
+    while(*thread_continue) {
+        void *p = malloc(block_size);
+        TEST_ASSERT_NOT_EQUAL(p, NULL);
+        free(p);
+    }
+}
+
+
+/** Test single malloc/free tracing
+ *
+ *  Given a memory trace mechanism
+ *  When perform single memory allocation/deallocation using malloc/free
+ *  Then tracing matches the operations
+ *
+ */
+static void test_case_single_malloc_free()
+{
+    const uint32_t num_op = 2;
     const size_t block_size = 126;
     const mem_trace_data_t *pmem = stats.op_data;
 
     test_clear_stats();
     mbed_mem_trace_set_callback(test_trace_cb);
+
     // Allocate a single memory block
     void *p = malloc(block_size);
     TEST_ASSERT_NOT_EQUAL(p, NULL);
+
     // Free the memory block
     free(p);
+
     // Stop tracing
     mbed_mem_trace_set_callback(NULL);
+
     // Check tracer result
-    check_sanity(2);
-    check_malloc_op(pmem ++, p, block_size);
+    check_sanity(num_op);
+    check_malloc_op(pmem++, p, block_size);
     check_free_op(pmem, p);
 }
 
-// Test all memory operations (malloc, realloc, free, calloc)
-static void test_case_all_memory_ops() {
+
+/** Test all memory operations (malloc, realloc, free, calloc) tracing
+ *
+ *  Given a memory trace mechanism
+ *  When perform all memory operations
+ *  Then tracing matches the operations
+ *
+ */
+static void test_case_all_memory_ops()
+{
+    const uint32_t num_op = 5;
     const size_t malloc_size = 40, realloc_size = 80, nmemb = 25, size = 10;
     const mem_trace_data_t *pmem = stats.op_data;
 
     test_clear_stats();
     mbed_mem_trace_set_callback(test_trace_cb);
+
     // Allocate a single memory block, the realloc it
     void *p_malloc = malloc(malloc_size);
     TEST_ASSERT_NOT_EQUAL(p_malloc, NULL);
     void *p_realloc = realloc(p_malloc, realloc_size);
     TEST_ASSERT_NOT_EQUAL(p_realloc, NULL);
+
     // Use calloc() now
     void *p_calloc = calloc(nmemb, size);
-    //TEST_ASSERT_NOT_EQUAL(p_calloc, NULL);
+    TEST_ASSERT_NOT_EQUAL(p_calloc, NULL);
+
     // Free the realloc() pointer first, then the calloc() one
     free(p_realloc);
     free(p_calloc);
+
     // Stop tracing
     mbed_mem_trace_set_callback(NULL);
+
     // Check tracer result
-    check_sanity(6);
-    check_malloc_op(pmem ++, p_malloc, malloc_size);
-    check_realloc_op(pmem ++, p_realloc, p_malloc, realloc_size);
-    // calloc() calls malloc() internally
-    check_malloc_op(pmem ++, p_calloc, nmemb * size);
-    check_calloc_op(pmem ++, p_calloc, nmemb, size);
-    check_free_op(pmem ++, p_realloc);
+    check_sanity(num_op);
```

Comment:
In the last version,  the argument to check_sanity() was 6, now changed to 5, is this something expected?

Reply:
It was 6 because `calloc` was traced twice as `calloc + malloc`
I was even worse when heap stats was enabled but this is fixed by #5614